build-push/bin/build-push.sh

```diff
@@ -0,0 +1,373 @@
+    # This will be verified, see check_dependencies()
+    NATIVE_GOARCH="${NATIVE_GOARCH:-$($RUNTIME info --format='{{.host.arch}}')}"
+    PARALLEL_JOBS="${PARALLEL_JOBS:-$($RUNTIME info --format='{{.host.cpus}}')}"
+
+    dbg "Found native go-arch: $NATIVE_GOARCH"
+    dbg "Found local CPU count: $PARALLEL_JOBS"
+
+    if [[ -z "$NATIVE_GOARCH" ]]; then
+        die_help "$E_NOGOARCH '$RUNTIME'"
+    elif ! type -P jq &>/dev/null; then
+        die_help "$E_NOJQ"
+    fi
+
+    # Not likely overridden, but keep the possibility open
+    PLATFORMOS="${PLATFORMOS:-linux}"
+
+    # Env. vars set by parse_args()
+    FQIN=""                   # required (fully-qualified-image-name)
+    CONTEXT=""                # required (directory path)
+    NOPUSH=1                  # optional (1 means push, 0 means do not)
+    ARCHES=("$NATIVE_GOARCH") # optional (Native architecture always first|only item)
+    TAGCMD=""                 # optional
+    MODCMD=""                 # optional
+    declare -a BUILD_ARGS
+    BUILD_ARGS=()             # optional
+    REGSERVER=""              # parsed out of FQIN
+    NAMESPACE=""              # parsed out of FQIN
+    IMGNAME=""                # parsed out of FQIN
+    unset NAMESPACE_USERNAME  # lookup based on $NAMESPACE when $NOPUSH=1
+    unset NAMESPACE_PASSWORD  # lookup based on $NAMESPACE when $NOPUSH=1
+
+    # List of variable names to export for --tagcmd and --modcmd
+    _CMD_ENV="RUNTIME PLATFORMOS FQIN CONTEXT NOPUSH ARCHES TAGCMD BUILD_ARGS REGSERVER NAMESPACE IMGNAME TAGCMDVAL BUILDIID"
+}
+
+parse_args() {
+    local -a args
+    local arg
+    local archarg
+    local -a archargs
+    local nsu_var
+    local nsp_var
+
+    dbg "in parse_args()"
+
+    if [[ $# -lt 2 ]]; then
+        die_help "$E_NOREQ"
+    fi
+
+    args=("$@")  # Special-case quoting: Will NOT separate quoted arguments
+    for arg in "${args[@]}"; do
+        dbg "Processing parameter '$arg'"
+        case "$arg" in
+            --arches=*)
+                archarg=$(tr ',' ' '<<<"${arg:9}")
+                if [[ -z "$archarg" ]]; then die_help "$E_ONEARCH '$arg'"; fi
+                archargs+=($archarg)
+                IFS=$'\n' ARCHES+=($(sort -u <<<"${archargs[*]}"))
+                ;;
+            --arches)
+                # Argument format not supported (to simplify parsing logic)
+                die_help "$E_ONEARCH '$arg'"
+                ;;
+            --tagcmd=*)
+                # Bash argument processing automatically strips any outside quotes
+                TAGCMD="${arg:9}"
+                ;;
+            --tagcmd)
+                die_help "$E_TAGVAL '$arg'"
+                ;;
+            --modcmd=*)
+                MODCMD="${arg:9}"
+                ;;
+            --modcmd)
+                die_help "$E_MODVAL '$arg'"
+                ;;
+            --nopush)
+                dbg "Nopush flag detected, will NOT push built images."
+                NOPUSH=0  # i.e. "Yes, do not push"
```

Comment:
> Nooooo.... please no

Oof...having to guess what you mean by that :smile:   I already re-implemented this to use only positive logic, storing `1` in `$PUSH` when the `--nopush` flag is given.  That's ***MUCH*** easier to read/understand.  However, even this could possibly be simplified further.

I think (tell me if I'm wrong), when CI is running on a branch and intended to push images, it will explicitly be given access to the secrets.   Otherwise I think it's a safe assumption that "no-password" == the caller just wants to build locally.  

Would that be a worthwhile further improvement?

Reply:
Clearification:  "no-password" meaning `$<NAMESPACE>_PASSWORD` is empty or unset.

Reply:
What I meant was, if you're going to use `NOPUSH` as a variable name, it should be set to 1 when `--nopush` is given. `NOPUSH=0` in the above context is Chaotic Evil.

I don't have a good enough sense for CI secrets management to reply to your question. Is it possible for secrets to silently go away? In that case the job would succeed without pushing, probably for a very long time without anyone noticing. If that's possible/likely, I think it might make sense to keep it as-is (fail loudly if secrets are missing). If it's unlikely, then I'm OK with your judgment call either way.